src/thinblock.cpp

```diff
@@ -247,6 +248,107 @@ bool CXThinBlock::CheckBlockHeader(const CBlockHeader &block, CValidationState &
     return true;
 }
 
+bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strCommand, unsigned nHops)
+{
+    if (!pfrom->ThinBlockCapable())
+    {
+        LOCK(cs_main);
+        Misbehaving(pfrom->GetId(), 100);
+        return error("%s message received from a non thinblock node, peer=%d", strCommand, pfrom->GetId());
+    }
+
+    int nSizeThinBlock = vRecv.size();
+    CXThinBlock thinBlock;
+    vRecv >> thinBlock;
+
+    // Message consistency checking
+    if (!IsThinBlockValid(pfrom, thinBlock.vMissingTx, thinBlock.header))
+    {
+        LOCK(cs_main);
+        Misbehaving(pfrom->GetId(), 100);
+        return error("Invalid %s received", strCommand);
+    }
+
+    // Is there a previous block or header to connect with?
+    {
+        LOCK(cs_main);
+        uint256 prevHash = thinBlock.header.hashPrevBlock;
```

Comment:
I'm concerned about all the cs_main locks we're aquiring and releasing and aquiring again...i think we do this about 3 times here.  While for thinblock processing it's not a big deal , for expedited it is important.  We can easily a 1/10th of second waiting for a cs_main lock.  I would look at consolidating them here so that we only take the lock once and hold it until we don't need it...so in checkng the POW (IsThinBlockValid), contextualcheckblockheader and also checking to see if the block connects...etc...all should be under one cs_main lock..rather than aquire/release/aquire/release...cs_main is recursive so you don't need worry about taking it twice, just do not release it until your done with it.